### ISSUE TITLE
Polyfill instance methods with extensions

### ISSUE THREAD

**Issue by Tyrrrz** [Sergio0694/PolySharp#6]
### Description (optional)

First off, awesome idea for the project. I was hoping someone would do it :)

I was wondering if you had any plans to polyfill certain built-in instance methods with extensions.
For example, these are some of the polyfills I have in my projects:

https://github.com/Tyrrrz/CliWrap/blob/ccc3087f0744dce17899a433591b3e50fd18e90b/CliWrap/Utils/Polyfills.Collections.cs#L1-L17

https://github.com/Tyrrrz/CliWrap/blob/ccc3087f0744dce17899a433591b3e50fd18e90b/CliWrap/Utils/Polyfills.Streams.cs#L1-L57

They allow me to seamlessly call methods only available on newer frameworks, while polyfilling the functionality on older frameworks.

### Rationale

Besides attributes, this could be an additional aspect of polyfilling that might be worth exploring. As a user, I can see great benefit in this as it would allow me to replace my own polyfills that I copy around between projects with a single package.

### Proposed API

I propose a similar approach that I've been using: extensions in a global namespace. This makes the polyfill callsites fully source-compatible with natural callsites.

### Drawbacks

This would require to figure out how to actually implement certain polyfills using existing functionality, and might not be trivial in some cases.

### Alternatives

Just writing polyfills manually.

### Other thoughts

None.

**Comment by menees** [comment]
I agree. This is a great project, and it would be nice if it also filled in some of the new .NET instance methods via extensions.

I use the [Microsoft.CodeAnalysis.NetAnalyzers](https://github.com/dotnet/roslyn-analyzers#microsoftcodeanalysisnetanalyzers) code analysis rules on projects that target both .NET Framework 4.8 and .NET 6.0. Rule [CA2249](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2249) recommends using `Contains(char)` instead of `IndexOf(char) >= 0`, which is great for .NET 6.0. But I have to add a `Contains(char)` polyfill for my .NET Framework targets to make this work. It would be great if Poly# provided implementations like these:

``` C#
/// <summary>
/// Returns a value indicating whether a specified character occurs within this string.
/// </summary>
/// <param name="text">The string to search in.</param>
/// <param name="value">The character to seek.</param>
/// <remarks>This method performs an ordinal (case-sensitive and culture-insensitive) comparison.</remarks>
/// <returns>true if the value parameter occurs within this string; otherwise, false.</returns>
public static bool Contains(this string text, char value)
	=> text.IndexOf(value) >= 0;

/// <summary>
/// Returns a value indicating whether a specified string occurs within this string, using the specified comparison rules.
/// </summary>
/// <param name="text">The string to search in.</param>
/// <param name="value">The string to seek.</param>
/// <param name="comparison">One of the enumeration values that specifies the rules to use in the comparison.</param>
/// <returns>true if the value parameter occurs within this string, or if value is the empty string (""); otherwise, false.</returns>
public static bool Contains(this string text, string value, StringComparison comparison)
	=> text.IndexOf(value, comparison) >= 0;

/// <summary>
/// Determines whether the end of the <paramref name="text"/> string instance matches the specified character.
/// </summary>
/// <param name="text">The text to check.</param>
/// <param name="ch">The character to compare to the character at the end of <paramref name="text"/>.</param>
/// <returns>
/// True if <paramref name="text"/> is non-empty and ends with <paramref name="ch"/>.
/// False if <paramref name="text"/> is null, empty, or doesn't end with <paramref name="ch"/>.
/// </returns>
public static bool EndsWith(this string text, char ch)
	=> text.Length > 0 && text[text.Length - 1] == ch;

/// <summary>
/// Determines whether the start of the <paramref name="text"/> string instance matches the specified character.
/// </summary>
/// <param name="text">The text to check.</param>
/// <param name="ch">The character to compare to the character at the start of <paramref name="text"/>.</param>
/// <returns>
/// True if <paramref name="text"/> is non-empty and starts with <paramref name="ch"/>.
/// False if <paramref name="text"/> is null, empty, or doesn't start with <paramref name="ch"/>.
/// </returns>
public static bool StartsWith(this string text, char ch)
	=> text.Length > 0 && text[0] == ch;
```

And here are some unit tests using [MSTest v2](https://github.com/microsoft/testfx) and [Shoudly](https://github.com/shouldly/shouldly):

``` C#
[TestMethod]
public void ContainsChar()
{
	string.Empty.Contains('x').ShouldBeFalse();
	"Test".Contains('e').ShouldBeTrue();
	"Test".Contains('S').ShouldBeFalse();
}

[TestMethod]
public void ContainsString()
{
	string.Empty.Contains("x", StringComparison.OrdinalIgnoreCase).ShouldBeFalse();
	"Test".Contains("es", StringComparison.OrdinalIgnoreCase).ShouldBeTrue();
	"Test".Contains("ES", StringComparison.OrdinalIgnoreCase).ShouldBeTrue();
}

[TestMethod]
public void EndsWith()
{
	string.Empty.EndsWith('x').ShouldBeFalse();
	"Test".EndsWith('t').ShouldBeTrue();
	"Test".EndsWith('T').ShouldBeFalse();
}

[TestMethod]
public void StartsWith()
{
	string.Empty.StartsWith('x').ShouldBeFalse();
	"Test".StartsWith('t').ShouldBeFalse();
	"Test".StartsWith('T').ShouldBeTrue();
}
```

**Comment by cremor** [comment]
Here are quite a few examples of which polyfill extension methods would be possible: https://github.com/SimonCropp/Polyfill#extensions

**Comment by Bouke** [comment]
My project contains also various polyfills to bridge the gap. I have compared https://github.com/Sergio0694/PolySharp and https://github.com/meziantou/Meziantou.Polyfill, and I've found that:
- Both didn't contain the extension methods that I have defined. That's no surprise as the gap in API differences between net48 and net8.0 is rather big.
- They overlap with the C# 12 features this project is filling, meaning I'd have to drop Polysharp.

What do people typically do here; go with the other projects? Define your own polyfills? Some other project that works in _addition_ to Polysharp?

**Comment by AvremelM** [comment]
@Sergio0694 Any chance we can get an official response here? 
This would be tremendously useful, if potentially a large expansion of scope.

**Pros**:
- Incredibly useful
- Makes sense for a polyfill-type package to include polyfill for apis like this
- Solves issues for users having to try to depend on both PolySharp and Polyfill or their own custom extensions
- Much of the work has already been done in other open source repos and can (likely) just be copied over

**Cons**:
- Broad expansion of this project's actual scope (even if not its conceptual scope)
- Other?

**Note**: 
This wouldn't completely obviate packages like @SimonCropp's excellent Polyfill package, which is source-only (not source-generated) and also adds new types like `Polyfills.Guard` for unpolyfillable static methods like `ArgumentException.ThrowIfNull` etc.

**Comment by SimonCropp** [comment]
@AvremelM in what scenario would you need to "depend on both PolySharp and Polyfill"?

**Comment by AvremelM** [comment]
@SimonCropp Purely hypothetical, assuming that someone was already using PolySharp but really wanted Polyfill's method pollyfills. But you're right, that would be pretty weird to depend on both and I have to imagine there would be conflicts if someone did (I haven't tried).
Also to clarify, that sentence should be understood as `depend on both ( PolySharp && (Polyfill || Custom || Other) )`. 

Also, I'm sorry if tagging you was bothersome, I forgot to consider that before I @'ed you out of the blue. 😅

**Comment by SimonCropp** [comment]
@AvremelM no worries about pinging me.

> assuming that someone was already using PolySharp but really wanted Polyfill's method pollyfills

i would hope this is never the case. Polyfill is designed to be a superset of PolySharp. So if there is anything missing in Polyfill that exists in PolySharp, then please let me know

**Comment by AvremelM** [comment]
@SimonCropp There are some differences between the two that may lead to people preferring PolySharp. The ability to opt in/out of specific polyfills, for instance. [See also](https://github.com/Sergio0694/PolySharp/issues/86#issuecomment-1781801731)

The only thing I've personally had issues with is the IDE ergonomics of source-only package vs source-generator, which are slightly different, at least for Rider. 
Polyfill files show up in Find in Files results, and "Go To Definition" actions open a physical file that is actually editable (!). Neither are ideal for me.

(Starting to veer a bit off topic here, but not sure where to move this to)

**Comment by Sergio0694** [comment]
> "Any chance we can get an official response here?"

This is not something I'm planning on adding, as it's outside the scope of PolySharp. This project is specifically meant to be for polyfills for **language features**. See other examples of similar requests: #119, #58, #49. So I'd like to keep things simpler here 🙂

Thank you!
